Detectors/EMCAL/reconstruction/src/StuDecoder.cxx

```diff
@@ -0,0 +1,241 @@
+      //      std::cout << "ERROR: Wrong payload size (=" << std::dec << payloadsize << ") for ShortPaload \n";
+      throw STUDecoderError(feeID, STUDecoderError::ErrorCode_t::SHORT_PAYLOAD_SIZE_UNEXPECTED);
+    }
+  }
+
+  if (mDebug >= 3) {
+    std::cout << "==== paylaod size ===\n";
+    int currentpos = 0;
+    while ((currentpos < buffer.size())) {
+      auto currentword = buffer[currentpos++];
+      std::cout << std::dec << (currentpos - 1) << ")   0x" << std::hex << currentword << std::endl;
+    }
+  }
+}
+
+void StuDecoder::decode()
+{
+
+  init();
+
+  auto& buffer = mRawReader.getPayload().getPayloadWords();
+
+  mCFGWord0 = buffer[0];
+  mCFGWord1 = buffer[1];
+  mL0mask = buffer[2];
+  mL1GammaHighThreshold = buffer[3];
+  mShortPayloadRate = buffer[4];
+  mL0bits = buffer[5];
+  mL1JetHighThreshold = buffer[6];
+  mL1GammaLowThreshold = buffer[9];
+  mL1JetLowThreshold = buffer[12];
+  mCFGword13 = buffer[13];
+  mRegionEnable = buffer[14];
+  mFrameReceived = buffer[15];
+  mCFGword16 = buffer[16];
+
+  // decode L1Jet High/Low indices
+  int offset = getCFGWords();
+  decodeL1JetPatchIndices(&buffer[offset]);
+
+  // decode L1Gamma High/Low indices
+  offset += (2 * getL1JetIndexWords() + getL0indexWords());
+  decodeL1GammaPatchIndices(&buffer[offset]);
+
+  if (!isFullPayload())
+    return;
+
+  // decode FastOR data
+  offset += (2 * getL1GammaIndexWords());
+  decodeFastOrADC(&buffer[offset]);
+
+  //  std::cout << "lastWord = 0x" << std::hex << buffer[offset + getRawWords()] << std::endl;
+  return;
+}
+
+void StuDecoder::decodeL1JetPatchIndices(const uint32_t* buffer)
+{
+  int offset = getL1JetIndexWords(); // offset for Jet Low threshold
+
+  if (mDebug >= 2) {
+    for (int i = 0; i < offset; i++)
+      std::cout << std::dec << i << ")  0x" << std::hex << buffer[i] << "   <- Jet-high-word\n";
+    for (int i = 0; i < offset; i++)
+      std::cout << std::dec << i << ")  0x" << std::hex << buffer[offset + i] << "   <- Jet-low-word\n";
+  }
+
+  int nSubregionEta = getSubregionsEta();
+  int nSubregionPhi = getSubregionsPhi();
+  bool isFired = false;
+
+  int jetSize = 2 + getParchSize(); // 2 for 2x2-patch, 4 for 4x4=patch
+
+  for (Int_t ix = 0; ix < nSubregionEta - (jetSize - 1); ix++) {
+    uint32_t row_JetHigh = buffer[ix];
+    uint32_t row_JetLow = buffer[ix + offset];
+    for (Int_t iy = 0; iy < nSubregionPhi - (jetSize - 1); iy++) {
+      isFired = (mSTU == STUtype_t::ESTU) ? (row_JetHigh & (1 << (nSubregionPhi - jetSize - iy))) : (row_JetHigh & (1 << iy));
+      if (isFired) {
+        mL1JetHighPatchIndex.push_back(((ix << 8) & 0xFF00) | (iy & 0xFF));
+      }
+      isFired = (mSTU == STUtype_t::ESTU) ? (row_JetLow & (1 << (nSubregionPhi - jetSize - iy))) : (row_JetLow & (1 << iy));
+      if (isFired) {
+        mL1JetLowPatchIndex.push_back(((ix << 8) & 0xFF00) | (iy & 0xFF));
+      }
+    }
+  }
+
+  return;
+}
+
+void StuDecoder::decodeL1GammaPatchIndices(const uint32_t* buffer)
```

Comment:
In the way I see in the code buffer is a sequence of 32 bit words with 4 8 bit words, a struct can be defined as

```c++
struct GammaPatchWord{
  union {
    uint32_t mData;
    struct {
       uint32_t mWord3: 8;
       uint32_t mWord2: 8;
       uint32_t mWord1: 8;
       uint32_t mWord0: 8;
    };
  };
};
```
See particularly as exampled how the different word types are defined for ITSMFT https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/ITSMFT/common/reconstruction/include/ITSMFTReconstruction/GBTWord.h and adapted in a similar way for FOCAL https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/FOCAL/reconstruction/include/FOCALReconstruction/PadWord.h

you would call the function via
```c++
decodeL1GammaPatchIndices(gsl::span<const GammaPatchWord>(reinterpret_cast<const GammaPatchWord *>(buffer + first, nwords)));
```